lib/strings.nix

```diff
@@ -324,6 +325,53 @@ rec {
   */
   escapeShellArgs = concatMapStringsSep " " escapeShellArg;
 
+  /* Test whether the given name is a valid POSIX shell variable name.
+
+     Type: string -> bool
+
+     Example:
+       isValidShellVar "foo_bar000"
+       => true
+       isValidShellVar "0-bad.jpg"
+       => false
+  */
+  isValidShellVar = name: match "[a-zA-Z_][a-zA-Z0-9_]*" name != null;
+
+  /* Make shell variables from Nix variables.
+
+     This function expects an attribute set of key-value pairs to be translated
+     into shell variables. If the values are strings (or string-coercible), then
+     the generated code is POSIX sh-compatible.
+
+     Values can also be lists of strings or attribute sets of strings, which are
+     translated into Bash-style arrays and associative arrays respectively.
+     Make sure that the shell for which you're writing supports this syntax.
+
+     Type: attrsOf (string | listOf string | attrsOf string) -> string
+
+     Example:
+       let
+         foo = "value";
+         bar = "value";
+       in ''
+         ${lib.mkShellVars { inherit foo bar; }}
+         [[ "$foo" == "$bar" ]]
+       ''
+  */
+  mkShellVars = vars: concatStringsSep "\n" (lib.mapAttrsToList (name: value:
```

Comment:
I wonder if perhaps there should be a well-known attribute for declaring a local variable? E.g., `vars._local`, when set the assignments would use `declare` and otherwise `declare -g`.

Should be noted that the current implementation would turn a Nix attribute set into a local shell variable while lists and atomic Nix values will be turned in global shell variables.

Reply:
Alternatively, maybe we always should define local variables? This would discourage the use of global variables and the user could always override the behavior by calling `declare -g {varname}` themselves.

Reply:
Ah, but there's a problem: the `declare` built-in is not specified by POSIX (and e.g. Dash or busybox `sh` don't understand `declare a=b`).

Since this should absolutely be POSIX compatible for string variables, and I still don't want to complicate the implementation, users will have to mark variables local explicitly. I think this is acceptable given that it should be a rare use case.